cpp/src/aztec/honk/composer/composer_helper/permutation_helper.hpp

```diff
@@ -1,118 +1,89 @@
-#include <stdint.h>
-#include <stddef.h>
-#include <vector>
+#pragma once
+
+#include <ecc/curves/bn254/fr.hpp>
+#include <polynomials/polynomial.hpp>
 #include <proof_system/proving_key/proving_key.hpp>
+
+#include <algorithm>
+#include <cstdint>
+#include <initializer_list>
+#include <cstdint>
+#include <cstddef>
+#include <utility>
+#include <vector>
+
 namespace honk {
-// Enum values spaced in increments of 30-bits (multiples of 2 ** 30).
-enum WireType { LEFT = 0U, RIGHT = (1U << 30U), OUTPUT = (1U << 31U), FOURTH = 0xc0000000 };
 
 /**
- * @brief cycle_node represents a particular witness at a particular gate. Used to collect permutation sets
- *
+ * @brief cycle_node represents the index of a value of the circuit.
+ * It will belong to a CyclicPermutation, such that all nodes in a CyclicPermutation
+ * must have the value.
+ * The total number of constaints is always <2^32 since that is the type used to represent variables, so we can save
+ * space by using a type smaller than size_t.
  */
 struct cycle_node {
+    uint32_t wire_index;
     uint32_t gate_index;
-    WireType wire_type;
-
-    cycle_node(const uint32_t a, const WireType b)
-        : gate_index(a)
-        , wire_type(b)
-    {}
-    cycle_node(const cycle_node& other)
-        : gate_index(other.gate_index)
-        , wire_type(other.wire_type)
-    {}
-    cycle_node(cycle_node&& other)
-        : gate_index(other.gate_index)
-        , wire_type(other.wire_type)
-    {}
-    cycle_node& operator=(const cycle_node& other)
-    {
-        gate_index = other.gate_index;
-        wire_type = other.wire_type;
-        return *this;
-    }
-    bool operator==(const cycle_node& other) const
-    {
-        return ((gate_index == other.gate_index) && (wire_type == other.wire_type));
-    }
 };
-typedef std::vector<std::vector<cycle_node>> CycleCollector;
+using CyclicPermutation = std::vector<cycle_node>;
 
 /**
- * Compute wire copy cycles
- *
- * First set all wire_copy_cycles corresponding to public_inputs to point to themselves.
- * Then go through all witnesses in w_l, w_r, w_o and w_4 (if program width is > 3) and
- * add them to cycles of their real indexes.
+ * Compute all CyclicPermutations of the circuit. Each CyclicPermutation represents the indices of the values in the
+ * witness wires that must have the same value.
  *
  * @tparam program_width Program width
  * */
 template <size_t program_width, typename CircuitConstructor>
-void compute_wire_copy_cycles(CircuitConstructor& circuit_constructor, CycleCollector& wire_copy_cycles)
+std::vector<CyclicPermutation> compute_wire_copy_cycles(const CircuitConstructor& circuit_constructor)
```

Comment:
Returning a value is cheap thanks to RVO (return value optimization) and makes the control flow clearer.